go/evm_test.go

```diff
@@ -0,0 +1,115 @@
+package evm
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"io/ioutil"
+	"math/big"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+type testCase struct {
+	Name string `json:"name"`
+	Hint string `json:"hint"`
+	Code code   `json:"code"`
+	Want want   `json:"expect"`
+}
+
+type code struct {
+	Bin string `json:"bin"`
+	Asm string `json:"asm"`
+}
+
+type want struct {
+	Stack   []hexBigInt `json:"stack"`
+	Success bool        `json:"success"`
+	Return  string      `json:"return"`
+}
+
+// A hexBigInt is a *big.Int that can be read from a JSON hex string.
+type hexBigInt struct {
+	*big.Int
+}
+
+// UnmarshalJSON unmarshals the buffer into i.Int; it expects the input to be
+// string-quoted.
+func (i *hexBigInt) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	if i.Int == nil {
+		i.Int = new(big.Int)
+	}
+	return i.Int.UnmarshalJSON([]byte(s))
+}
+
+// StackInts returns the underlying *big.Int values of w.Stack, unwrapping them
+// from within the JSON-unmarshalling helper.
+func (w *want) StackInts() []*big.Int {
+	b := make([]*big.Int, len(w.Stack))
+	for i, s := range w.Stack {
+		b[i] = s.Int
+	}
+	return b
+}
+
+func TestEVM(t *testing.T) {
+	var tests []testCase
+	t.Run("setup", func(t *testing.T) {
+		const testSrc = "../evm.json"
+		buf, err := ioutil.ReadFile(testSrc)
+		if err != nil {
+			fatalAndBugReport(t, "ioutil.ReadFile(%q) error %v", testSrc, err)
+		}
+
+		if err := json.Unmarshal(buf, &tests); err != nil {
+			fatalAndBugReport(t, "json.Unmarshal([contents of %q], %T) error %v", testSrc, &tests, err)
+		}
```

Comment:
`ioutil.ReadFile` is deprecated since 1.16, the new function is `os.ReadFile`. But instead of reading and copying file content inside the buffer, you could just open a file descriptor and directly decode the content to `tests` variable.

```go
file, err := os.Open("../evm.json")
if err != nil {
    fatalAndBugReport(t, "ioutil.ReadFile(%q) error %v", testSrc, err)
}

if err := json.NewDecoder(file).Decode(&tests); err != nil {
    fatalAndBugReport(t, "json.Unmarshal([contents of %q], %T) error %v", testSrc, &tests, err)
}
```

Reply:
TIL about `json.NewDecoder()`! Thank you, I'll update it in the morning.